microcosm_eventsource/controllers.py

```diff
@@ -0,0 +1,134 @@
+"""
+Event controllers.
+
+"""
+from inflection import camelize
+
+from microcosm_flask.conventions.crud_adapter import CRUDStoreAdapter
+from microcosm_flask.conventions.encoding import with_context
+from microcosm_flask.naming import name_for
+from microcosm_flask.operations import Operation
+from microcosm_pubsub.conventions import created
+from werkzeug.exceptions import UnprocessableEntity
+
+from microcosm_eventsource.errors import (
+    IllegalInitialStateError,
+    IllegalStateTransitionError,
+)
+
+
+class EventController(CRUDStoreAdapter):
+
+    @property
+    def sns_producer(self):
+        return self.graph.sns_producer
+
+    def create(self, event_type, **kwargs):
+        """
+        Create an event, validating the underlying state machine.
+
+        """
+        self.validate_required_fields(event_type, **kwargs)
+        parent = self.store.retrieve_most_recent(**kwargs)
+        version, state = self.process_state_transition(parent, event_type)
+        event = self.create_event(parent, event_type, version, state, **kwargs)
+        self.publish_event(event)
+        return event
+
+    def validate_required_fields(self, event_type, **kwargs):
+        """
+        Validate type-specific required fields.
```

Comment:
Inspect the required fields for event types and raise 422 before the DB catches the constraint violation (and raises 500)